File: aoc-2021/day-01/day_1.hs
-- stack --resolver lts-16.27 script 

main :: IO ()
main = do
  content <- readFile "input.txt"
  let numbers = map read $ lines content
  putStrLn $ "Part 1: " ++ show (countIncreasing numbers)
  putStrLn $ "Part 2: " ++ show (countIncreasing $ movingSum numbers)

movingSum :: [Int] -> [Int]
movingSum (x:[]) = [x]
movingSum x = f : r
  where f = sum $ take 3 x
        r = movingSum $ tail x

countIncreasing :: [Int] -> Int
countIncreasing x = sum $ map comp $ zip x (tail x) 
  where comp (x,y) 
          | x < y = 1
          | otherwise = 0

File: aoc-2021/day-01/day_1.py
from toolz import sliding_window

with open("input.txt") as f:
    numbers = list(map(int, f.read().splitlines()))


def pair_compare(*x):
    return x[0] < x[1]


print(
    "Part 1:",
    sum(
        map(
            pair_compare,
            sliding_window(2, numbers),
        )
    ),
)
print(
    "Part 2:",
    sum(
        map(
            pair_compare,
            sliding_window(
                2, map(lambda z: sum(z) / len(z), sliding_window(3, numbers))
            ),
        )
    ),
)
